app/lib/actions/candlepin/environment/set_content.rb

```diff
@@ -19,20 +19,46 @@ class SetContent < Candlepin::Abstract
           param :content_ids, Array
         end
 
-        def run
-          saved_cp_ids = ::Katello::Resources::Candlepin::Environment.
+        def existing_ids
+          ::Katello::Resources::Candlepin::Environment.
               find(input[:cp_environment_id])[:environmentContent].map do |content|
             content[:contentId]
           end
+        end
+
+        def run
+          saved_cp_ids = existing_ids
           add_ids    = input[:content_ids] - saved_cp_ids
           delete_ids = saved_cp_ids - input[:content_ids]
-
-          output[:add_ids]         =    add_ids
-          unless add_ids.empty?
-            output[:add_response]    = ::Katello::Resources::Candlepin::Environment.
-              add_content(input[:cp_environment_id], add_ids)
+          retries = 0
+          max_retries = 2
+          completed = false
+          until completed
+            output[:add_ids]         =    add_ids
+            unless add_ids.empty?
```

Comment:
I'd probably simplify this according to the ruby style guide:

https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals
https://github.com/bbatsov/ruby-style-guide#loop-with-break